### PR TITLE
Fix some optimization warnings in jit.

### PR DESCRIPTION
**Pull request by sandreenko**
Fix inefficient x64 struct layouts and passings by value.
It gives a tiny IR (-0.03%) and memory consumption (-0.09%) improvement.

That is an optimization follow-up from the Friday learning day, thanks to PVS-Studio.

Correctness fixes will be later.

### CODE REVIEW

**Comment by sandreenko** [comment]
PTAL @dotnet/jit-contrib

File: src/jit/compiler.h [hunk 1]
```diff
@@ -193,9 +193,9 @@ inline HfaElemKind HfaElemKindFromType(var_types type)
 typedef const char* VarName; // Actual ASCII string
 struct VarScopeDsc
 {
+    unsigned  vsdVarNum;  // (remapped) LclVarDsc number
```
Comment (mikedn) [inline src/jit/compiler.h]: What does moving this do?

Comment (sandreenko) [reply src/jit/compiler.h]: thanks for catching it, it was strange formatting. 
That was a debug only warning because of 8 bytes `VarName vsdName;`.
I will move it 3 lines down (after `vsdVarNum`) so it won't look that awkward.

It doesn't affect release, but cleans the warning in my report, so I don't need to look at it each time and it is a good thing for the checked build.

File: src/jit/compiler.h [hunk 2]
```diff
@@ -7206,9 +7206,9 @@ class Compiler
 
     struct IPmappingDsc
     {
+        emitLocation  ipmdNativeLoc; // the emitter location of the native code corresponding to the IL offset
```
Comment (BruceForstall) [inline src/jit/compiler.h]: Couldn't this go after the "next" pointer and have the same packing/alignment?

File: src/jit/emit.h [hunk 3]
```diff
@@ -255,8 +255,6 @@ struct insGroup
     UNATIVE_OFFSET igNum;     // for ordering (and display) purposes
     UNATIVE_OFFSET igOffs;    // offset of this group within method
     unsigned int   igFuncIdx; // Which function/funclet does this belong to? (Index into Compiler::compFuncInfos array.)
-    unsigned short igFlags;   // see IGF_xxx below
```
Comment (BruceForstall) [inline src/jit/emit.h]: Why did you move these two (igFlags/igSize)? They should already be appropriately aligned/packed on all architectures.

File: src/jit/jitconfig.h [hunk 4]
```diff
@@ -21,11 +21,11 @@ class JitConfigValues
             MethodName* m_next;
             int         m_methodNameStart;
             int         m_methodNameLen;
-            bool        m_methodNameWildcardAtEnd;
             int         m_classNameStart;
             int         m_classNameLen;
-            bool        m_classNameWildcardAtEnd;
             int         m_numArgs;
+            bool        m_methodNameWildcardAtEnd;
```
Comment (BruceForstall) [inline src/jit/jitconfig.h]: You should add a comment here that these are here to improve packing. Note that improving packing in this struct is not very important...

File: src/jit/loopcloning.h [hunk 5]
```diff
@@ -327,8 +327,8 @@ struct LC_Ident
         Null,
     };
 
-    unsigned  constant; // The constant value if this node is of type "Const", or the lcl num if "Var"
     LC_Array  arrLen;   // The LC_Array if the type is "ArrLen"
+    unsigned  constant; // The constant value if this node is of type "Const", or the lcl num if "Var"
```
Comment (BruceForstall) [inline src/jit/loopcloning.h]: Seems like `LC_Array` itself could be rearranged to put its pointer first.

File: src/jit/emit.h (hunk 3 above)
Comment (BruceForstall) [reply src/jit/emit.h]: Note that `emitterStaticStats()` is one way to help figure out if the packing is tight. (Change
```
#define EMITTER_STATS 0       // Collect stats on the emitter.                       
```
to define to `1` and rebuild (in release), then run to see its output.

File: src/jit/loopcloning.h (hunk 5 above)
Comment (sandreenko) [reply src/jit/loopcloning.h]: Yes, but it won't change its `size/8`.

File: src/jit/compiler.h (hunk 2 above)
Comment (sandreenko) [reply src/jit/compiler.h]: It won't matter for x64, but a good rule is to sort them by size in descending order (`emitLocation ` has max size).
Do you want me to move it after `next`

File: src/jit/emit.h (hunk 3 above)
Comment (sandreenko) [reply src/jit/emit.h]: The change affects only `defined(DEBUG) || defined(LATE_DISASM)`.

Note: in my opinion `/d1reportSingleClassLayoutinsGroup` is more useful, you can rebuild only clrjit project from VS and see the result much faster.

File: src/jit/compiler.h (hunk 2 above)
Comment (BruceForstall) [reply src/jit/compiler.h]: IMO, "next" pointers being first makes more semantic sense. I think it's perhaps faster on some architectures in some cases for something to be at offset zero.

File: src/jit/emit.h (hunk 3 above)
Comment (BruceForstall) [reply src/jit/emit.h]: I don't care about DEBUG. And `LATE_DISASM` is also not enabled by default. I didn't know about the switch you give; can you show sample output?

**Comment by CarolEidt** [comment]
> Correctness fixes will be later.

Can you elaborate? Were there also correctness issues identified by PVS-Studio?

**Comment by sandreenko** [comment]
> Can you elaborate? Were there also correctness issues identified by PVS-Studio?

Yes, but they need more time to analyze and find the right fixes.

File: src/jit/emit.h (hunk 3 above)
Comment (sandreenko) [reply src/jit/emit.h]: There are two VS switches `/d1reportSingleClassLayout<name>` and `/d1reportAllClassLayout`.
Add them to CoreCLr.sln->clrjit->properties->C/C++ -> Command Line -> Additional Options:
image

and see the layout of the chosen struct when you build clrjit:
image

**Comment by sandreenko** [comment]
@BruceForstall I have updated the PR, please take a look.

File: src/jit/compiler.h (hunk 2 above)
Comment (BruceForstall) [reply src/jit/compiler.h]: > a good rule is to sort them by size in descending order

Technically, you want to sort them by alignment requirement, in descending order. If a large struct has alignment of 1 (such as for a struct of all `char` types), you wouldn't want to put it before the pointers, with an alignment requirement of `sizeof(void*)`
